File: lingting-core/src/main/java/live/lingting/framework/system/CommandResult.java
package live.lingting.framework.system;

import live.lingting.framework.util.StreamUtils;
import lombok.Getter;
import lombok.SneakyThrows;

import java.io.IOException;
import java.io.InputStream;
import java.nio.file.Files;
import java.time.Duration;

/**
 * @author lingting 2022/6/25 12:01
 */
@Getter
public class CommandResult {

	protected final Command command;

	protected final int exitCode;

	protected final long end;

	protected final Duration duration;

	protected String stdOut;

	protected String stdErr;

	public CommandResult(Command command, int exitCode) {
		this.command = command;
		this.exitCode = exitCode;
		this.end = System.currentTimeMillis();
		this.duration = Duration.ofMillis(end - command.startTime);
	}

	@SneakyThrows
	public synchronized String getStdOut() {
		if (stdOut == null) {
			InputStream stream = stdOut();
			stdOut = new String(StreamUtils.read(stream), command.charset);
		}
		return stdOut;
	}

	@SneakyThrows
	public synchronized String getStdErr() {
		if (stdErr == null) {
			InputStream stream = stdErr();
			stdErr = new String(StreamUtils.read(stream), command.charset);
		}
		return stdErr;
	}

	public InputStream stdOut() throws IOException {
		return Files.newInputStream(command.stdOut.toPath());
	}

	public InputStream stdErr() throws IOException {
		return Files.newInputStream(command.stdErr.toPath());
	}

	public void clean() {
		command.clean();
	}

}

File: lingting-core/src/main/kotlin/live/lingting/framework/system/CommandResult.kt
package live.lingting.framework.system

import java.io.Closeable
import java.io.InputStream
import java.nio.charset.Charset
import java.time.LocalDateTime
import live.lingting.framework.time.DateTime

/**
 * @author lingting 2025/2/20 13:52
 */
open class CommandResult @JvmOverloads constructor(
    protected val command: Command,
    val pid: Long,
    val startTime: LocalDateTime = command.startTime,
    val endTime: LocalDateTime = DateTime.current(),
    val code: Int,
    val charset: Charset = command.charset,
    private val outPipe: CommandPipe,
    private val errPipe: CommandPipe,
) : Closeable {

    open fun outStream(): InputStream = outPipe.stream()

    open fun errStream(): InputStream = errPipe.stream()

    open fun outBytes(): ByteArray = outStream().use { it.readAllBytes() }

    open fun errBytes(): ByteArray = errStream().use { it.readAllBytes() }

    open fun outString(): String = outBytes().toString(charset)

    open fun errString(): String = errBytes().toString(charset)

    override fun close() {
        command.close()
    }

}
